Sources/NIO/MulticastChannel.swift

```diff
@@ -0,0 +1,123 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// A `MulticastChannel` is a `Channel` that supports IP multicast operations: that is, a channel that can join multicast
+/// groups.
+///
+/// - note: As with `Channel`, all operations on a `MulticastChannel` are thread-safe.
+public protocol MulticastChannel: Channel {
+    /// Request that the `MulticastChannel` join the multicast group given by `group`.
+    ///
+    /// - parameters:
+    ///     - group: The IP address corresponding to the relevant multicast group.
+    ///     - promise: The `EventLoopPromise` that will be notified once the operation is complete, or
+    ///         `nil` if you are not interested in the result of the operation.
+    func joinGroup(_ group: SocketAddress, promise: EventLoopPromise<Void>?)
+
+    /// Request that the `MulticastChannel` join the multicast group given by `group` on the interface
+    /// given by `interface`.
+    ///
+    /// - parameters:
+    ///     - group: The IP address corresponding to the relevant multicast group.
+    ///     - interface: The interface on which to join the given group, or `nil` to allow the kernel to choose.
+    ///     - promise: The `EventLoopPromise` that will be notified once the operation is complete, or
+    ///         `nil` if you are not interested in the result of the operation.
+    func joinGroup(_ group: SocketAddress, interface: NIONetworkInterface?, promise: EventLoopPromise<Void>?)
+
+    /// Request that the `MulticastChannel` join the multicast group given by `group`.
+    ///
+    /// - parameters:
+    ///     - group: The IP address corresponding to the relevant multicast group.
+    /// - returns: An `EventLoopFuture` that will be notified once the operation is complete.
+    func joinGroup(_ group: SocketAddress) -> EventLoopFuture<Void>
+
+    /// Request that the `MulticastChannel` join the multicast group given by `group` on the interface
+    /// given by `interface`.
+    ///
+    /// - parameters:
+    ///     - group: The IP address corresponding to the relevant multicast group.
+    ///     - interface: The interface on which to join the given group, or `nil` to allow the kernel to choose.
+    /// - returns: An `EventLoopFuture` that will be notified once the operation is complete.
+    func joinGroup(_ group: SocketAddress, interface: NIONetworkInterface?) -> EventLoopFuture<Void>
```

Comment:
shouldn't this be _only_ in the proto extension? There's only one good implementation for that